src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertyAndValueDeserializer.cs

```diff
@@ -1022,7 +1023,31 @@ in resourceState.PropertyAndAnnotationCollector.GetCustomPropertyAnnotations(pro
             resourceState.PropertyAndAnnotationCollector.CheckForDuplicatePropertyNames(property);
             ODataResourceBase resource = resourceState.Resource;
             Debug.Assert(resource != null, "resource != null");
-            resource.Properties = resource.Properties.ConcatToReadOnlyEnumerable("Properties", property);
+
+            // To avoid repeated computations for the property verification logic inside the
+            // resource.Properties setter, we update the resource.Properties in-place
+            // without invoking the setter each time.
+            // Since the resource and properties are created internally by the reader,
+            // we don't need to verify the properties if we can guarantee that we are only
+            // adding properties with acceptable values (i.e. not ODataResourceValue).
+            // We use debug-time assertions instead to catch such bugs in tests.
+
+            // We should rethink the current approach to property verification.
+            // See: https://github.com/OData/odata.net/issues/3263
+
+            Debug.Assert(!(property.Value is ODataResourceValue), Strings.ODataResource_PropertyValueCannotBeODataResourceValue(property.Name));

```

Comment:
Would `Contract.Assert` be a better choice?
That aside, I did investigate when `ODataJsonPropertyAndValueDeserializer.ReadResourceValue[Async]` is executed and found it's during reading of `ODataError` and annotations where resource values are allowed. Specifically I identified that the following tests do trigger the `ReadResourceValue[Async]` method:
- MultipleTypeCustomInstanceAnnotationsOnErrorShouldRoundtrip
- ResourceCustomInstanceAnnotationOnErrorShouldRoundtrip
- ReadTopLevelErrorWithResourceValueInstanceAnnotation_Works
- ReadTopLevelErrorWithCollectionResourceValueInstanceAnnotation_Works
- ReadStreamPrimitivePropertyWithoutValueButWithInstanceAnnotations

Reply:
Thanks @gathogojr  for the additional extensive checks. I went with `Debug.Assert` with the intention that these checks would be stripped out of release builds.